### DISCUSSION TITLE
Implementing short-lived branches

### DISCUSSION THREAD

**Discussion by seporterfield** [Monstarrrr/rebutify#72] [Ideas & Suggestions]
## Long-lived branches are bottlenecks.

Encouraging long-lived branches has the following effects on us developers:

## 1. Starving main of fixes

We naturally tend to introduce the quality of life changes we need into our bottleneck branch, which starves main and any new branches. Everyone working on new branches is waiting for the bottleneck branch to merge.

Some current examples of this are (switching from feat/1_api to a new branch off main):
-  devs have to reinstall crispy-forms to have a non-broken build. That means devs have to install a package we got rid of ages ago and don't use, and take care to write their code in a way that doesn't break the outdated code, just to build their feature.
- devs have to deal with the broken prettier precommit hook, commits without --no-verify don't work. Effectively we don't have pre-commit working.
- eslint for nextjs is broken on main because that change is stuck in feat/1_api

## 2. Starving main of changes for features 

In a project where branches have a small scope, devs can contribute faster, because they don't need to wait as long if they have changes to implement that rely on other changes in an existing branch. Soon, that branch will be merged and they can piggyback off of it.

## 3. Merge hell

The older a branch, the harder it is to merge. More time is spent on solving merge conflicts than implementing changes.

## 4. Slow feedback loop

Having a dev environment where we can automatically deploy our trunk branch gives us feedback on the state of our application.

Feedback becomes a slow drip, or non-existent, when we use long-lived branches.

There are many more reasons to start using trunk based development (and get away from using long-lived branches), but these are the ones that I see affecting us the most now or in the future.

Wording for #3 and #4 taken from [this article](https://ardalis.com/trunk-based-development-vs-long-lived-feature-branches/)
It goes through the pros and cons of each approach.
The references there are a great read as well.

## What we can do
If we decide to not use long-lived branches, we can try the following strategies:

- use branches for small changes, even if they lead up to a big feature/issue
- tie branches to subtasks of issues, but don't delete the issue after merge
- set expiration dates for branches

Of course this doesn't mean pushing broken builds to main. We can use [feature flags](https://m.youtube.com/watch?v=AJa2B-twtG4) in extreme cases, but most of the time it should be possible to push changes without breaking the build.

This might mean implementing backend and frontend parts of features separately, but in small chunks.

## Conclusion

Long lived branches are going to hurt us in the long run, and lead to bad habits like starving main of good changes or necessary fixes.
There are enough strategies that allow us to work without long lived branches that we will have a far easier time transitioning away from these bottleneck branches than the trouble it's worth to go through merge hell, among other things.

Let's discuss concrete steps moving away from long lived branches, whether these are the ones suggested in this post or an idea you want to try.

**Comment by Monstarrrr** [comment]
## The conditions for reliable long-lived branches

It seems to me that long-lived branches **don't** cause **merge hell** if:  
- There are frequent merges of ``main`` into that branch  
- There is only one long-lived branch at a time  

And when it comes to starving ``main`` of features/fixes, it **isn't** an issue if:  
- Fixes/features impacting main and not just the long-lived branch are pushed separately to the feature branch (enforced during code reviews)  

I can see it becoming **increasingly difficult** to fulfill those criteria as the project evolve.  

## Pros & Cons of each approach  

We looked at the cons of long-lived branches but let's look at the pros & cons of both approaches.  

|  Approach  | Cons                              | Pros | 
|-------------|-----------------------------------|------|
| Trunk-based  | Relies on strong testing <br /> Not ideal for large changes<br /> Feature Flag* Complexity |  Quick feedback loop<br /> Reduced Merge Hell<br /> More collaboration<br /> Faster deployments for testing<br /> Easier for CI/CD |  
| Long-lived | Merge Hell<br /> Slow feedback loop<br /> Reduced collaboration<br /> Requires merging from main frequently | Isolation<br /> Code stability on main<br /> Good for big changes | 

*[What is a Feature Flag](https://launchdarkly.com/blog/what-are-feature-flags/)  

I do think we should move to a trunk-based development, but if we agree on this, then the question is **when** ?  
- After merging ``feat/1_auth`` ?
- After releasing the MVP ?

**Reply by seporterfield** [reply]
Nitpick: Approach names are swapped in the table

Thanks, very measured response

**Reply by seporterfield** [reply]
I'm for merging ``feat/1_auth`` right away.

Finishing the feature and working trunk-based otherwise with ``feat/1_auth`` until merge as an exception would be a fine compromise.

**Reply by Monstarrrr** [reply]
Merging now would impact #1 's content, and **may** introduce other type of extra work (not sure, but based on the fact that we've linked and structured issues & PRs I feel like it may).  

So I think the compromise is fine.  

Thoughts, @purple-void ?

**Reply by rebutify-developer** [reply]
We found that trunk-based means "All branches extend from one trunk/main branch, usually called the master branch." We're fine with working this way

[trunk-based vs feature-based](https://www.perforce.com/blog/vcs/trunk-based-development-or-feature-driven-development)
